### PR TITLE
KEYCLOAK-15955 Implement a pluggable Upgrade strategy

### PR DESCRIPTION
**Pull request by vmuzikar**
## JIRA ID
KEYCLOAK-15955

## Checklist:
- [X] Automated Tests
- [X] [Documentation](https://github.com/keycloak/keycloak-documentation) changes if necessary:
   https://github.com/keycloak/keycloak-documentation/pull/1062
- [ ] (If required) Discussed on [Keycloak DEV](https://groups.google.com/forum/#!forum/keycloak-dev)

## Additional Notes 
I'll add the docs once I've got the confirmation this is the right approach. ;)

### CODE REVIEW

**Comment by vmuzikar** [comment]
@slaskawi I was not able to run `make code/gen`. It was failing on:
```
operator-sdk generate k8s
INFO[0000] Running deepcopy code-generation for Custom Resource group versions: [keycloak:[v1alpha1], ] 
F1102 17:22:17.672959   12469 deepcopy.go:813] Hit an unsupported type invalid type for []invalid type
make: *** [code/gen] Error 255
```
It's apparently caused by `operator-sdk generate k8s` and it's happening for me on master too, so I assumed it's not caused by my changes.

However, I was able to run the rest of the commands from that make target, so hopefully the CRD is updated correctly.

**Comment by slaskawi** [comment]
@vmuzikar The CRD generation works fine on my laptop:

```
$ make code/gen
operator-sdk generate k8s
INFO[0000] Running deepcopy code-generation for Custom Resource group versions: [keycloak:[v1alpha1], ] 
INFO[0008] Code-generation complete.                    
operator-sdk generate crds --crd-version v1beta1
INFO[0000] Running CRD generator.                       
INFO[0000] CRD generation complete.                     
# This is a copy-paste part of `operator-sdk generate openapi` command (suggested by the manual)
which ./bin/openapi-gen > /dev/null || go build -o ./bin/openapi-gen k8s.io/kube-openapi/cmd/openapi-gen
./bin/openapi-gen --logtostderr=true -o "" -i ./pkg/apis/keycloak/v1alpha1 -O zz_generated.openapi -p ./pkg/apis/keycloak/v1alpha1 -h ./hack/boilerplate.go.txt -r "-"
API rule violation: list_type_missing,github.com/keycloak/keycloak-operator/pkg/apis/keycloak/v1alpha1,KeycloakRealmSpec,RealmOverrides

$ operator-sdk version  
operator-sdk version: "v0.18.2", commit: "f059b5e17447b0bbcef50846859519340c17ffad", kubernetes version: "v1.18.2", go version: "go1.13.10 linux/amd64"
```

**Review by slaskawi — CHANGES_REQUESTED** [review]
The code looks nice, but needs some changes:
- Please make sure Operator SDK works on your box. You may either build it from source or use a Fedora Docker Container. You may also find my email on our team's mailing list on a Docker container with prod utils.
- Please fix linter errors.
- You might simplify the code a bit.

File: pkg/controller/keycloak/keycloak_migrations.go
Comment (vmuzikar) [reply pkg/controller/keycloak/keycloak_migrations.go:60]: What the JIRA should be about? :) I mean it works, so it's not a bug. So, should it be a refactor task? Won't this migrator be deprecated soon anyway?

File: pkg/controller/keycloak/keycloak_migrations.go
Comment (vmuzikar) [reply pkg/controller/keycloak/keycloak_migrations.go]: Done.

File: pkg/controller/keycloak/keycloak_migrations_test.go
Comment (vmuzikar) [reply pkg/controller/keycloak/keycloak_migrations_test.go]: Done.

**Comment by vmuzikar** [comment]
@slaskawi Thanks for your review! I've pushed the requested changes. Didn't squash them yet so it's more easily apparent.

> The code looks nice, but needs some changes:
> 
> * Please make sure Operator SDK works on your box. You may either build it from source or use a Fedora Docker Container. You may also find my email on our team's mailing list on a Docker container with prod utils.

The Operator SDK seem to be working for me, it's just `generate k8s` doesn't. I'll investigate more why is it so but I think re-generating k8s is not required for this PR, correct?
> * Please fix linter errors.

Done.
> * You might simplify the code a bit.

Done.

**Comment by vmuzikar** [comment]
Fixed the `generate k8s`. I didn't have `GOROOT` set which was not a problem ... until now. https://github.com/operator-framework/operator-sdk/issues/1854#issuecomment-613202510 saved me. :)

File: pkg/controller/keycloak/keycloak_migrations.go
Comment (slaskawi) [reply pkg/controller/keycloak/keycloak_migrations.go:52]: oooh yeah.. sorry I misread this.

**Review by slaskawi — DISMISSED** [review]
The code looks fine. Could you please squash everything into a single commit?

**Comment by vmuzikar** [comment]
@slaskawi Ok, squashed.

Do we need to update the docs? Do we even have docs for the operator? :D Couldn't find any.

**Comment by jeesmon** [comment]
@vmuzikar Operator documentation seems to be from here: https://github.com/keycloak/keycloak-documentation/tree/master/server_installation/topics/operator

**Comment by slaskawi** [comment]
Dear contributor,

I'm putting this Pull Request on hold until we get our CI back (see https://github.com/keycloak/keycloak-operator/pull/266). Once the Github Actions Pull Request is merged, I will kindly ask you for rebasing your changes. Then, we'll proceed with a normal code review process.

Please see [this Keycloak DEV thread](https://groups.google.com/g/keycloak-dev/c/ijI9grpRw74) to learn more why we are migrating to Github Actions.

Sorry for inconvenience

**Comment by slaskawi** [comment]
@vmuzikar Could you please rebase this?

**Comment by slaskawi** [comment]
@vmuzikar Can I ask you to rebase this?

**Comment by vmuzikar** [comment]
> @vmuzikar Can I ask you to rebase this?

Done.

Working on the docs now.

**Comment by vmuzikar** [comment]
Docs PR: https://github.com/keycloak/keycloak-documentation/pull/1062

**Review by slaskawi — APPROVED** [review]

**Comment by slaskawi** [comment]
Integrated, thanks @vmuzikar !

**Comment by slaskawi** [comment]
@vmuzikar Could you also create a backport?

**Pull request merged**
